src/recoil/dotplot.ts

```diff
@@ -0,0 +1,197 @@
+import { atom, selector } from "recoil";
+import { currentOntologyState, dagDataStructureState } from ".";
+import { geneDataState } from "./genes";
+import { OntologyVertexDatum } from "../types/graph";
+
+import { scaleLinear } from "d3-scale";
+
+// is the dotplot drawer open?
+export const dotplotIsOpenState = atom<boolean>({
+  key: "dotplotIsOpen",
+  default: false,
+});
+
+export const rowHighlightedState = atom<string | null>({
+  key: "rowHighlighted",
+  default: null,
+});
+
+// how many celltypes max, in the dotplot?
+export const dotplotRenderThresholdState = atom<number>({
+  key: "dotplotRenderThreshold",
+  default: 200,
+});
+
+export const dotSizeMaxRadiusState = atom<number>({
+  key: "dotSize",
+  default: 10,
+});
+
+export const dotSizeMinRadiusState = atom<number>({
+  key: "dotSizeMin",
+  default: 1,
+});
+
+// are we currently under the max render threshold?
+export const dotplotEnabledState = selector<boolean>({
+  key: "dotplotEnabled",
+  get: ({ get }) => {
+    const dagDataStructure = get(dagDataStructureState);
+    const dotplotRenderThreshold = get(dotplotRenderThresholdState);
+
+    if (!dagDataStructure) {
+      return false;
+    } else {
+      return dagDataStructure.nodes.length < dotplotRenderThreshold;
+    }
+  },
+});
+
+// rows in the dotplot (celltypes), and their associated data
+export const dotplotRowState = selector<OntologyVertexDatum[] | null>({
+  key: "dotplotRows",
+  get: ({ get }) => {
+    const dagDataStructure = get(dagDataStructureState);
+    const dotplotEnabled = get(dotplotEnabledState);
+
+    if (dagDataStructure && dagDataStructure.nodes && dotplotEnabled) {
+      return dagDataStructure.nodes;
+    } else {
+      return null;
+    }
+  },
+});
+
+// columns in the dotplot (genes), as an array of names in format eg., 'ENSG00000196549'
+export const diffexpGenesDotplotState = selector<Array<any>>({
+  key: "diffexpGenesDotplot",
+  get: ({ get }) => {
+    const rows = get(dotplotRowState);
+    const ontology = get(currentOntologyState);
+
+    if (rows && ontology) {
+      const allDiffexpGenes = rows.map((row) => {
+        return ontology?.get(row.id)?.genes;
+      });
+
+      // flatten the array of arrays, remove duplicates and undefined values
+      const diffexpGenes = Array.from(new Set(allDiffexpGenes.flat().filter((genes) => genes !== undefined)));
```

Comment:
I'm assuming Set preserves order of the source array?

Reply:
Yes https://stackoverflow.com/questions/55460303/why-does-js-keep-insertion-order-in-set#:~:text=Yes%2C%20theoretically%2C%20sets%20are%20not,objects%20we%20use%20in%20programming.